src/EditorFeatures/Core.Wpf/InlineErrors/InlineErrorTag.cs

```diff
@@ -0,0 +1,126 @@
+using System.Diagnostics;
+using System.Windows;
+using System.Windows.Controls;
+using System.Windows.Documents;
+using System.Windows.Media;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editor.Implementation.Adornments;
+using Microsoft.VisualStudio.Text.Adornments;
+using Microsoft.VisualStudio.Text.Classification;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Formatting;
+
+namespace Microsoft.CodeAnalysis.Editor.InlineErrors
+{
+    internal class InlineErrorTag : GraphicsTag
+    {
+        public readonly string ErrorType;
+        private readonly DiagnosticData _diagnostic;
+
+        public InlineErrorTag(string errorType, DiagnosticData diagnostic, IEditorFormatMap editorFormatMap)
+            : base(editorFormatMap)
+        {
+            ErrorType = errorType;
+            _diagnostic = diagnostic;
+        }
+
+        /// <summary>
+        /// Creates a GraphicsResult object which is the error block based on the geometry and formatting set for the item.
+        /// </summary>
+        public override GraphicsResult GetGraphics(IWpfTextView view, Geometry bounds, TextFormattingRunProperties format)
+        {
+            var block = new TextBlock
+            {
+                FontFamily = format.Typeface.FontFamily,
+                FontSize = 0.75 * format.FontRenderingEmSize,
+                FontStyle = FontStyles.Normal,
+                Foreground = format.ForegroundBrush,
+                Padding = new Thickness(left: 2, top: 0, right: 2, bottom: 0),
+                VerticalAlignment = VerticalAlignment.Center,
+            };
+
+            var id = new Run(_diagnostic.Id);
+            var link = new Hyperlink(id)
+            {
+                NavigateUri = new Uri(_diagnostic.HelpLink)
+            };
+
+            link.RequestNavigate += HandleRequestNavigate;
+            block.Inlines.Add(link);
+            block.Inlines.Add(": " + _diagnostic.Message);
+            block.Measure(new Size(double.PositiveInfinity, double.PositiveInfinity));
+            block.Arrange(new Rect(block.DesiredSize));
+
+            var border = new Border
+            {
+                Background = format.BackgroundBrush,
+                Child = block,
+                CornerRadius = new CornerRadius(2),
+                // Highlighting lines are 2px buffer. So shift us up by one from the bottom so we feel centered between them.
+                Margin = new Thickness(0, top: 0, 0, bottom: 2),
+                Padding = new Thickness(2)
+            };
+
+            border.Measure(new Size(double.PositiveInfinity, double.PositiveInfinity));
+
+            view.ViewportWidthChanged += ViewportWidthChangedHandler;
+            // Need to set these properties to avoid unnecessary reformatting because some dependancy properties
+            // affect layout.
+            // TODO: Not sure if these are needed anymore since the errors are not intratextadornment tags
```

Comment:
Since this is an optimization to reduce reformatting likely worth keeping. It sets the properties to follow the textview, so reformatting will happen closer to that requirement. It's likely not needed because the reformatting of the errors won't interrupt text viewing, but still not bad to have.